### ISSUE TITLE
0.4.9 release

### ISSUE THREAD

**Issue by Bertram25** [OpenDungeons/OpenDungeons#9]
This version is aiming at providing a basic yet functional game with a few things to do, at least:

**General:**
- [x] The game is stable and doesn't crash at exit.
- [x] The game does use threads only when needed. (Through sf::Thread)
- [x] The code has been reviewed, is following conventions and doesn't use not GPL'ed code.
  (Still a few files to review.) #91
- [x] The license of every files is clear and defined in the https://github.com/OpenDungeons/OpenDungeons/blob/development/CREDITS file. #93
- [x] Fix CMake basic invocation flags handling - Done: #97 #42 #57 #58  

**General game logic added and free of main bugs:**
- [x] Spawn pools. Each faction has got his/her own configurable set of spawnable creatures.
- [x] Optimize pickup/dig mouse handling speed #103
- [x] Fix crash when a level creature has the same name of a later created one: #98
- [x] Possible crash on missile spawn? #120 
- [x] Ghost tile due to a bug in tile fullness server messages? #51
- [x] Double-click pick/drop makes the game crash #122 
- [x] The creature stats format upgrade #151 
- [x] No more lacking creatures stats #152 #245 
- [x] Creatures are spawned based on the available rooms and their active spots number and it is configurable. #257 
- [x] Better balancing through the use of a global balance sheet #153 (Sheet done, balancing is left to be done once @Danimal696 has got a rolling release with those features in.)

**Basic rooms added and functional**
- [x] Treasure room
  - [x] New Ground texture
  - [x] New Gold stacks
- [x] Portal (Lacks new model?) -> Let's not wait for this. @Danimal696 will come with something in time.
- [x] Temple (Lacks new model?) -> Let's not wait for this.
- [x] Training Room
  - [x] Class & file rename (RoomDojo -> RoomTrainingHall) #83 
  - [x] Better training dummy model
- [x] Dormitory
  - [x] The bed room objects shouldn't be deleted and recreated each time the room is changed.
  - [x] Afterward, check that when a creature dies, its bed is removed.
  - [x] Class & file rename (RoomQuarters -> RoomDormitory) #83 
- [x] Hatchery & food system #13
  - [x] All models integrated and the room can placed in game. #106
  - [x] The room spawns chickens wandering within the room limits.
  - [x] Creatures can eat them visually and calm their hunger.
- [x] Forge (Not actually useful for the next release)
- [x] Library  (Not actually useful for the next release)
- [x] First treasury tile owned should cost 0 to avoid being stuck - #87 
- [x] The rooms active spots aren't always complete #141 
- [x] Add support to destroy a room using a new button. Destroying a room gives back half the tile price. #108
- ~~Add support to disable a room for a specific player in a specific level #86~~

**Functional creature behavior:** 
- [x] Creature hunger system #13 
- [x] Creature behaving correctly in narrow passages (1-tile wide). Part of #4  
- [x] Separate water/lava and flying collision handling. #101
- [x] Fix creature behavior broken because of the maneuver action spam #104
- [x] Better creature fighting behaviour #154
- [x] Make workers flee on first enemy blow. #90 

**Basic Menus functional:**
- [x] Game mode and editor mode functional.
- [x] Minimap functional.
- [x] The basic menus are functional and look definite. #6
- [x] The remote clients shouldn't have to select a level. #70 
- [x] Show room pricing - First part of #89 
- [x] Add a button to grab workers  and fighters easily in the menu #84 #85 

**Basic AI:** See #15
- [x] The AI can build every useful room.
- [x] The AI takes care of building rooms to get the maximum "active spots" (represented by room models).
- [x] The AI can drop fighters near-by other enemy creatures to trigger some fun.
- [x] AI should take care of having at least one treasury tile before searching for gold: After #87 is done.

**Graphics & content:**
- [x] All models are displayed fine even if not yet using normals. Done: #76 #100  #129 
- [x] Missing music/sounds: #95 #23 

**Content Bonus done:**
- Lan support
- Fixed and configurable alliances per level support.
- Factions (Keeper/Hero) and AI / Human player are configurable in a lobby before playing!!
- Improved GUI, graphics, models. New beds! ;)
- 3 functional traps that can be fought. Ranged traps now deal damage when they hit, not when they shoot at target.
- Functional editor with a loading menu
- Creatures can drop gold when digging it if they have already too much of it and no treasury room is reachable. Dug gold can be collected and put back to the treasury by either a worker or the Keeper itself when it is on a claimed tile.
- Creature definitions are overrideable in levels to create special creatures, or bosses.
- The equipment definition is now defined in the config/equipment.cfg file.
- Creatures can spawn with default equipment.
- Rooms and traps behaviour are now defined in config/rooms.cfg and config/traps.cfg files.
- Replay mode! You can watch your previous games. Still mainly aimed at developers, you'll be warned.
- Creatures can be slapped to death. And more importantly, those creepy chickens can be slapped while on a claimed tile

**Comment by hwoarangmy** [comment]
> Spawn pools. Each team has got a configurable set of creature spawnable.

The spawn pool is now linked to the seat which is unique per player. That means that 2 players in the same team can have different spawn pools.

**Comment by Bertram25** [comment]
I'll update the line. :)

**Comment by Bertram25** [comment]
According to this last bullet:
- Workers searching successfully for far away ground tile, walls to claim, gold or dirt to dig once their near-by counterparts are done. (Still depends on other improvement effects.)

@hwoarangmy Since you remade something like 70% of the pathfinding algorithm, do you know whether workers are now able to seek far away gold tiles to dig?
(I know that they are still restrained to what they see when no gold is marked)

@ everyone:
Do you think we could/should add support to make them seek for far away claimable tiles when they are idle for too long?

**Comment by hwoarangmy** [comment]
No, they only work where they see. Pathfinding works now and they would be able to go anywhere accessible if asked.
FYI, I've tried lately to run the 2v2 small map doing nothing for something like 2 hours. When I came back, all the gold in the map was mined. The AI ally had even dug my walls to get the gold near my dungeon temple ^^
All the tiles in the map were claimed.

> Do you think we could/should add support to make them seek for far away claimable tiles when they are idle for too long?

I think that's the kind of thing that can be done later, when we will have a real game :)

**Comment by Bertram25** [comment]
> I think that's the kind of thing that can be done later, when we will have a real game :)

Ok. I'll open an issue for later. In the meanwhile, it's already quite fine. :)

**Comment by akien-mga** [comment]
Wow we're really close to the 0.4.9 release :-)

**Comment by Bertram25** [comment]
Yeah, the main thing left now CMake stuff is mostly done is the creature def format. :)

(Disclaimer for below: This is not a complain, in fact, I'm quite happy about it.)
Note that @hwoarangmy is in Chuck Norris mode and he is already working on 0.6.3. ;)

**Comment by hwoarangmy** [comment]
^^
Indeed, I've prepared a huge PR for you to review :)
Many (too many ?) fixes. I had no time yesterday to do a PR as my ill younger daughter did not let me. The good point is that I had time to test and, so far, it seems pretty stable. I know that release time is not the best to do such a PR but I think it is worth it. And I'm sure our crazy tester @akien-mga will make sure I've not forgotten anything ^^

**Comment by Bertram25** [comment]
Arf, don't even think you have the right to apologize for accelerating the development pace of this project by ten. ;P

> And I'm sure our crazy tester @akien-mga will make sure I've not forgotten anything ^^

Yeah, your foolish efforts to make Akien the Archlich kneel before the testing charge are worthless, for he'll wipe all your patches with effortless testing reports. ;)

I also hope your daughter is feeling better, and that is wasn't that bad.

**Comment by hwoarangmy** [comment]
> I also hope your daughter is feeling better, and that is wasn't that bad.

We will see tonight if she can sleep :)

**Comment by Danimal696** [comment]
And the list keeps on growing :) , next 0.4.9 release will be something nice

**Comment by Bertram25** [comment]
AI apart, we're close! Do you guys want to keep the AI for that release or should we move it out for now? (Seems nobody is convinced by jumping into AI coding, I guess until we have a rock-solid core game experience.)

**Comment by akien-mga** [comment]
I think the AI could be moved to 0.5.0, it will be a big task, so I think it wouldn't hurt to have a stable release to show to the world before that, and one that is already great for LAN games :-)

**Comment by Bertram25** [comment]
> so I think it wouldn't hurt to have a stable release to show to the world before that

"A release to show to the world". It's always a daunting sentence. O.O' ;)

**Comment by hwoarangmy** [comment]
Well, if we want to have "A release to show to the world", I think we need to have an AI that at least builds basic rooms and can attack. For now on, if you have no gold into your dungeon temple, it will not even try to reach you...

But in 0.4.9, there is "Add support to disable a room for a specific player in a specific level" which can be set to later IMHO.

**Comment by Bertram25** [comment]
> But in 0.4.9, there is "Add support to disable a room for a specific player in a specific level" which can be set to later IMHO.

Really? I thought we wanted to add that to be able to disable the library and forge building (since they do nothing and cost money) Or maybe you want to be able to show them anyway?

**Comment by akien-mga** [comment]
> Well, if we want to have "A release to show to the world", I think we need to have an AI that at least builds basic rooms and can attack. For now on, if you have no gold into your dungeon temple, it will not even try to reach you...

It's true, but I still think that it would be way better to release 0.4.9 in its current state than to still have 0.4.8 as our latest "showcase" release. We can make it clear that 0.4.9 is not full game, that it's already interesting to try but probably not to spend hours playing it against the AI.

**Comment by akien-mga** [comment]
> Really? I thought we wanted to add that to be able to disable the library and forge building (since they do nothing and cost money) Or maybe you want to be able to show them anyway?

I'd leave them for now, because it's great to see them :-) Maybe we can just add to their price tooltip a mention that those rooms are not useful currently?

**Comment by Bertram25** [comment]
> I'd leave them for now, because it's great to see them :-) Maybe we can just add to their price tooltip a mention that those rooms are not useful currently?

I'd leave a disclaimer in news about the release anyway. :)
Ok, let's put #86 out. :)

**Comment by Bertram25** [comment]
Btw, thanks to whoever completed the summary. It'll be very useful to create the release news. :)

**Comment by hwoarangmy** [comment]
> Btw, thanks to whoever completed the summary. It'll be very useful to create the release news. :)

I thought it was you ^^

**Comment by Bertram25** [comment]
> > Btw, thanks to whoever completed the summary. It'll be very useful to create the release news. :)
> 
> I thought it was you ^^

Mostly, but not only. :)

**Comment by Bertram25** [comment]
Ok, what's left is start balancing and improving the ai. :)
I'd propose once we've got better balancing, to play a bit together and see how it turns ;)

**Comment by Danimal696** [comment]
We are sooo close!

**Comment by Bertram25** [comment]
Release TODO List almost complete!! Some balancing is left!! :O 

I'd propose to divide the auto-healing rate of all creature by 5 as a first step, btw. :)

**Comment by hwoarangmy** [comment]
> I'd propose to divide the auto-healing rate of all creature by 5 as a first step, btw. :)

True, they heal too fast making sleeping almost useless...

**Comment by Danimal696** [comment]
seconded, someone should also get some spawn conditions for heroes

**Comment by Bertram25** [comment]
Done for both in latest PR. :)

**Comment by akien-mga** [comment]
Well well, it looks like the last step to the release is a slightly better balancing.

@Bertram25 Maybe a (last?) pre-release Windows build could be nice to let @Danimal696 and others experiment with the AI and further tweak the config?

I can feel 0.4.9 coming :-D BTW we should start tagging issues with 0.5.0 instead of always "Later" depending on their priority.

**Comment by hwoarangmy** [comment]
Yes, 0.4.9 is near ^^
But some balancing is definitely needed as currently, when you play a game and creatures get to level 10, battles takes ages (and I'm not even talking on the fact that now, the AI saves its creatures so it is even worst :) )

**Comment by Bertram25** [comment]
> @Bertram25 Maybe a (last?) pre-release Windows build could be nice to let @Danimal696 and others experiment with the AI and further tweak the config?

Yes, I'd like to finish another PR first, though. But once it is done, yes, definitely. :) I'll take the opportunity to fix #275 at the same time if nobody else does it.

> I can feel 0.4.9 coming :-D BTW we should start tagging issues with 0.5.0 instead of always "Later" depending on their priority.

Well, you have a point, but I don't want us to start tagging stuff for the next release before we have settled what we want in it. Thus, may I request we open a 0.5.0 release features discussion on the forums first? Once we have a common vision of it, I'll be ok to add the 0.5.0 milestone. That said, I'd like everyone to stay focused on 0.4.9 until release (aka in bugfix mode) so we finally put something out asap! :D

**Comment by Bertram25** [comment]
> But some balancing is definitely needed as currently, when you play a game and creatures get to level 10, battles takes ages (and I'm not even talking on the fact that now, the AI saves its creatures so it is even worst :) )

Eh eh. But AFAICT, lowering the heal per turn has already improved the thing already. It's all about atk vs def depending on the level I guess?

**Comment by Bertram25** [comment]
Ah, and I wondered, should the creature received a minimal damage value whatever its defense is, so that a big creature encountering 50 low level minions at least "gets hurt" in the process?

**Comment by hwoarangmy** [comment]
> Eh eh. But AFAICT, lowering the heal per turn has already improved the thing already. It's all about atk vs def depending on the level I guess?

Well, I didn't had the opportunity to test that until now. I've rebased yesterday but I didn't test any combat. I will see that tonight ^^

> Thus, may I request we open a 0.5.0 release features discussion on the forums first?

There is already a thread for that:
http://forum.freegamedev.net/viewtopic.php?f=35&t=5821
Even if it will probably need to be adapted, it is a good starting point ^^

**Comment by Bertram25** [comment]
Yeah, i wrote there. Thanks for the reminder. If all goes well, I should have a PR for all my assigned issue at the beginning of next week.

**Comment by Bertram25** [comment]
Ok, once the PR about the boulder trap is in, it will be time for the first release candidate used to do balancing. :D :yum:

**Comment by Bertram25** [comment]
I consider the game ready for a first RC, made for balancing. If we can merge the hand before that and fix the XP table problem, it would be cool but IMHO, it shouldn't prevent a new rolling release. Thus, I'll do it as soon as I've time. ;)

**Comment by hwoarangmy** [comment]
If I am not mistaken, everything is ready ^^

**Comment by Bertram25** [comment]
> If I am not mistaken, everything is ready ^^

Almost. I'm on removing the useless files. And Akien wanted to add the manpage, too AFAIK.

**Comment by akien-mga** [comment]
Woops, the release is waiting for me! I'll do the manpage asap, I guess Debian has some nice documentation on how to create one :-)

**Comment by akien-mga** [comment]
There were scattered comments in PRs about disabling (some?) console commands for the release. Do you think we should do it, or release 0.4.9 as is?

IMO we don't need to bother hiding too many dev features for now, because:
- This release is still an alpha meant to show the world how great OD is becoming, and people won't be surprised by cheating debug commands in an alpha version IMO.
- The multiplayer feature is currently limited to playing with acquaintances (you can't see available servers and connect to a random one currently), so most likely people will play with friends that they trust to be fair players; if not, we can let them handle that :-P

WDYT?

BTW should we put some order in our current levels set before the release? e.g. remove some test maps, remove the enemy traps and creatures next to seat 1 in the 2v2 map, etc.?

**Comment by hwoarangmy** [comment]
Concerning console, I think we should let it because as you say, people will only play with friends as there is no server. So, there is not much interest into cheating.
And it can help us to debug if needed

**Comment by Bertram25** [comment]
As for levels, this has been discussed already and we decided at the time that the test level should remain for this version to show the engine is being tested (as we're in alpha.).
I added support for the map description to separate more clearly test maps from actual ones.
So, I'd request that we don't bother with that for 0.4.9.

On 0.5.0, on the other hand, we'll have to disable the display of test levels and maybe hide certain features such as console commands unless in dev mode.
If you guys agree with it, I'll open an issue.

**Comment by Bertram25** [comment]
@Danimal696 Btw, the RC1 being imminent IMHO, are you ok to handle the official news about it, or do you guys want to handle that in a more common way maybe?
(This issue summary was meant to be used as a reminder for the changelog.)

**Comment by akien-mga** [comment]
> As for levels, this has been discussed already and we decided at the time that the test level should remain for this version to show the engine is being tested (as we're in alpha.).

Sounds good :-) I just thought that the 2v2 map is a nice one to actually play OD both in single and multiplayer (even if it's really basic), so I'd like to ship a "plain" 2v2 map (no enemy tiles or monsters or traps next to seat 1 basically).

Remember that's it's the one we've played the most in our LANs :-p

**Comment by Bertram25** [comment]
> Remember that's it's the one we've played the most in our LANs :-p

You definitely have a point. ;) But isn't the multiplayer version of this map, free of testing content? I just can't remember now...

When seeing the beast, it is definitely true that the most needed thing is maps now... Should we add one or two small maps as suggested?

**Comment by akien-mga** [comment]
> You definitely have a point. ;) But isn't the multiplayer version of this map, free of testing content? I just can't remember now...

Yes most likely. Let's leave it like that for 0.4.9, and as you said we can add a feature to separate test levels from "release" levels. Maybe two different tabs in the GUI? I think it's still relevant to leave test levels in releases as long as OD is in alpha.

**Comment by hwoarangmy** [comment]
> no enemy tiles or monsters or traps next to seat 1 basically

Yep, as @Bertram25 noticed, the multiplayer map don't have the test features. The only difference between the 4 players is a construct worker for player 1. It should probably be removed.

Concerning test maps, as already said, in alpha phase I wouldn't bother about it. Moreover, we don't have many maps so the list would fill... empty. So, yes, it would be cool to have 1 or 2 more real maps...

**Comment by Danimal696** [comment]
Yes, i can handle the news, having pics will depend on Elvano, since he is the only with those permits and... seems MIA. But besides that ill post it first on forum before the official release on homepage to do some joint refinement.

**Comment by akien-mga** [comment]
Elvano is still hanging around on #opendungeons on Freenode, so I can grab him there if need be :-)

**Comment by Danimal696** [comment]
Good! please telll him to be ready for this :D !

**Comment by Bertram25** [comment]
> So, yes, it would be cool to have 1 or 2 more real maps...

Ok, I've got ideas for a 1vs1 and a 1vs2 map. I'll add them tonight. I will also need a clear consensus about the HLSL shaders and the directX renderer so we can kill all that asap. :)

**Comment by akien-mga** [comment]
Seems we'll release 0.4.9 two years and a half after its development began: https://github.com/OpenDungeons/OpenDungeons/commit/8e48136b36a335997bdcedec1e847b2716d3a831
Kudos to @oyvindln for this commit :-P

**Comment by hwoarangmy** [comment]
> HLSL shaders and the directX renderer

Concerning shaders, you know better than me what to do and I will follow your judgement (and try your patch ;) )
Concerning DirectX, as I said in another thread, it can be removed IMHO if there are no regressions for anybody testing (ie @Bertram25 , @Danimal696 and me as I guess we are the only ones concerned)

**Comment by akien-mga** [comment]
And... 0.4.9 milestone is achieved :-)

**Comment by Bertram25** [comment]
\O/ I'll merge dev into stable and create a '0.4.9' tag there.
Note to self: Push also to gitorious.org

I'll create a release on github.com and will have a go at doing a win32 zip as well. :)
Once all that is done, it will be time to do some news!! :D

**Comment by Danimal696** [comment]
About time!

**Comment by Bertram25** [comment]
RC1 Snapshot out! :)
I'll add the link to the Win32 binary tomorrow. Need some sleep. ;)

**Comment by Bertram25** [comment]
Once my PR is in, I'll redo the RC1 tag as the fixes are simple yet nice enough to be part of it.

**Comment by akien-mga** [comment]
@Bertram25 Elvano has set us both as "Developer+" on the sourceforge project so that we can manage the Files section. So feel free to upload your Windows binary there once it's ready.

**Comment by Danimal696** [comment]
and while you are at it, please update the pic and the video

**Comment by akien-mga** [comment]
Which video should be used? The boring one I posted on http://forum.freegamedev.net/viewtopic.php?f=36&t=6050 ? :-)

**Comment by Danimal696** [comment]
any is better than the 3 years old one

**Comment by hwoarangmy** [comment]
> Which video should be used? The boring one I posted

Why don't you cut the boring parts ?

**Comment by akien-mga** [comment]
> Why don't you cut the boring parts ?

Because I'm not good with video editing ;-) I just meant "boring" as "watching Akien play against the IA for 15 minutes is not the most entertaining". We need a real trailer with a good pace, max 3 min or so.

**Comment by Bertram25** [comment]
For now, I'd say let's upload this one. I wonder whether qubodup would agree making one for us once the release is publicly out? He's usually good at that.
Ah and ChemBroTron made a very good video for VT. Maybe he'll agree to do one for OD?
http://www.youtube.com/channel/UCDs0Jp8WYKMcPvVcHcYuPsA

**Comment by Bertram25** [comment]
As the news draft is ready and all that's lacking a website back online, I'll close this issue as I consider the 0.4.9 version ready. If any blocking fixes are to be added, we'll do a 0.4.10. :)

**Comment by akien-mga** [comment]
@Bertram25 Do you know whether the current Win32 snapshot for 0.4.9 RC1 is the one to use for 0.4.9 final? The only commit that might be lacking in it would be 000453e50 (I'm not sure whether it's in or not, it probably depends on whether you made the binary from your branch or from the 0.4 branch).

I'll update the source tarball with the latest 0.4 branch too.

Elvano has posted the release announcement: http://opendungeons.sourceforge.net/

**Comment by akien-mga** [comment]
Actually the PR #391 might be missing too, it was merged afterwards (but maybe you included it yourself in the binary).

**Comment by akien-mga** [comment]
For now I uploaded the final 0.4.9 source tarball (based on the current 0.4 branch) and renamed the 2014-12-22 Win32 zipfile to say that it's the final 0.4.9 version.

@Bertram25 Can you tag 0.4.9 in github?

**Comment by Bertram25** [comment]
Hey, :D 

It is soo cool to see the release online!

> Actually the PR #391 might be missing too, it was merged afterwards (but maybe you included it yourself in the binary).

Yeah, the portal fix was put in as a last minute addition.

> For now I uploaded the final 0.4.9 source tarball (based on the current 0.4 branch) and renamed the 2014-12-22 Win32 zipfile to say that it's the final 0.4.9 version.

Thanks, IMHO, there is nothing more to do concerning the Windows version, as no changes is actually needed to enjoy the 0.4.9 experience there.

> @Bertram25 Can you tag 0.4.9 in github?

Sure, asap. :)

**Comment by Bertram25** [comment]
N.B.: The 0.4.9 tag will actually point on the same commit than the 0.4.9-RC1 one.

**Comment by akien-mga** [comment]
Great. I'm correcting some typos and formatting issues in the release announcement, and will send the diff to Elvano.

**Comment by Bertram25** [comment]
@akien-mga Ah, just in case, don't forget to tell him to update the windows and linux download links! :)
And sorry for the typos. :S

**Comment by Danimal696** [comment]
Isnt it time to open a new 0.5 release Issue? ;)

**Comment by Bertram25** [comment]
> Isnt it time to open a new 0.5 release Issue? ;)

It is. I'm on it. :)
